File: selenium-java/src/test/java/com/example/tests/steps/RobotActionsLoadSteps.java
package com.example.tests.steps;

import com.example.tests.config.DriverHolder;
import com.example.tests.pages.RobotActionsHomePage;
import com.example.tests.support.ElementHandler;
import com.example.tests.support.WaitHandlers;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

import java.net.URI;
import java.util.List;
import java.util.logging.Logger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Step definitions for {@code features/robotactions-load.feature}.
 *
 * Key design decisions:
 * <ul>
 *   <li>Uses {@link DriverHolder#get()} (ThreadLocal) — consistent with Hooks pattern.</li>
 *   <li>{@link ScenarioContext} (also ThreadLocal) carries per-scenario state.</li>
 *   <li>{@link #assertNoConsoleSevereErrors()} is wrapped in try/catch because
 *       {@code driver.manage().logs().get(LogType.BROWSER)} throws on Chrome 132+
 *       when browser logs are not available.  The step logs a warning and continues
 *       rather than failing on a capabilities gap.</li>
 * </ul>
 */
public class RobotActionsLoadSteps {

    private static final Logger LOG = Logger.getLogger(RobotActionsLoadSteps.class.getName());
    private static final String RA_HOME = "https://robotactions.com";

    // ── Helpers ───────────────────────────────────────────────────────────

    private WebDriver driver() {
        return DriverHolder.get();
    }

    private RobotActionsHomePage page() {
        return new RobotActionsHomePage(driver());
    }

    /**
     * Wait tiers for the steps that assert on driver state (URL, title) rather
     * than on an element the page object owns. Same instance the page objects
     * use — see {@link WaitHandlers}.
     */
    private WaitHandlers waits() {
        return WaitHandlers.forDriver(driver());
    }

    /** Element interaction for the few steps that touch the DOM directly. */
    private ElementHandler elements() {
        return ElementHandler.forDriver(driver());
    }

    // ── Background ────────────────────────────────────────────────────────

    @Given("I open the RobotActions home page")
    public void openHomePage() {
        driver().get(RA_HOME);
    }

    /** Generic "go to URL" for negative tests (404 page, unknown hash). */
    @When("I navigate to path {string}")
    public void navigateToPath(String path) {
        driver().get(RA_HOME + path);
    }

    /** Assertion used by negative tests + deep-link sanity checks. */
    @Then("the page title should contain {string}")
    public void pageTitleShouldContain(String text) {
        // Wait briefly for SPA hydration so the title settles.
        waits().awaitTrue(d -> d.getTitle() != null && d.getTitle().contains(text),
            waits().shortWait());
        String actual = driver().getTitle();
        assertTrue(actual.contains(text),
            "page title '" + actual + "' does not contain '" + text + "'");
    }

    @Given("I wait for the SPA to hydrate")
    public void waitForSpaHydration() {
        page().waitForHydration();
    }

    // ── Hero + primary CTAs ───────────────────────────────────────────────

    /**
     * Asserts that the current H1 text contains at least one of the three
     * rotating hero phrases.  A fixed-string assertion would be intermittently
     * flaky because the phrases cycle every ~20 s.
     */
    @Then("the hero heading should contain one of the known phrases")
    public void heroHeadingContainsKnownPhrase() {
        assertTrue(page().heroHeadingMatchesKnownPhrase(),
            "hero heading '" + page().heroHeadingText()
            + "' does not match any of the known phrases: "
            + RobotActionsHomePage.KNOWN_HERO_PHRASES);
    }

    /**
     * Asserts that the stable tagline beneath the rotating H1 is visible.
     * "No credit card required • Cancel anytime • 30-day free trial" never rotates.
     */
    @Then("the stable hero tagline should be visible")
    public void stableHeroTaglineVisible() {
        assertTrue(page().isStableTaglineVisible(),
            "stable hero tagline containing 'No credit card required' is not visible");
    }

    /** Legacy step kept for compatibility — delegates to the rotating-phrase check. */
    @Then("the hero heading should contain {string}")
    public void heroHeadingContains(String fragment) {
        String text = page().heroHeadingText();
        assertTrue(text.contains(fragment),
            "hero heading text '" + text + "' does not contain '" + fragment + "'");
    }

    @Then("I should see the {string} button")
    public void shouldSeeButton(String label) {
        assertTrue(page().isButtonVisible(label),
            "button '" + label + "' is not visible");
    }

    @Then("I should see the heading {string}")
    public void shouldSeeHeading(String label) {
        assertTrue(page().isHeadingVisible(label),
            "heading '" + label + "' is not visible");
    }

    // ── Top-level nav: hash anchors ───────────────────────────────────────

    @When("I click the {string} nav link")
    public void clickNavLink(String label) {
        page().clickNavLink(label);
    }

    @Then("the URL fragment should be {string}")
    public void urlFragmentShouldBe(String fragment) {
        waits().awaitTrue(d -> d.getCurrentUrl().endsWith(fragment), waits().shortWait());
        assertTrue(driver().getCurrentUrl().endsWith(fragment),
            "expected URL to end with '" + fragment + "' but was '"
            + driver().getCurrentUrl() + "'");
    }

    @Then("the section with id {string} should be in viewport")
    public void sectionInViewport(String id) {
        assertTrue(page().isSectionInViewport(id),
            "section #" + id + " is not in the viewport");
    }

    // ── Dropdowns ─────────────────────────────────────────────────────────

    @When("I click the {string} nav button")
    public void clickNavButton(String label) {
        page().clickNavButton(label);
    }

    @When("I click the {string} button")
    public void clickButton(String label) {
        page().clickButton(label);
    }

    @Then("the {word} dropdown should be visible")
    public void dropdownShouldBeVisible(String dropdownName) {
        // Sentinel map lives on the page object — steps don't re-declare it.
        String sentinel = RobotActionsHomePage.dropdownSentinel(dropdownName);
        assertTrue(page().isDropdownLinkVisible(sentinel),
            "dropdown '" + dropdownName + "' does not appear to be open "
            + "(sentinel link '" + sentinel + "' is not visible)");
    }

    /**
     * Verifies that every label in the data table is visible as a link inside
     * the (already-open) dropdown.  Uses {@code contains()} matching so labels
     * like "AI QA Agent" match the DOM text "AI QA AgentNEW".
     *
     * The feature file's table has a header row "label" — Cucumber-JVM 7.x
     * includes the header in {@code dataTable.asList()}, so we skip element 0.
     */
    @Then("the {word} dropdown should contain the items:")
    public void dropdownContainsItems(String dropdownName, DataTable dataTable) {
        List<String> rows = dataTable.asList();
        List<String> labels = rows.subList(1, rows.size());
        for (String label : labels) {
            assertTrue(page().isDropdownLinkVisible(label),
                "dropdown '" + dropdownName + "' item '" + label + "' is not visible");
        }
    }

    @When("I click the {string} dropdown item")
    public void clickDropdownItem(String label) {
        page().clickDropdownItem(label);
    }

    @Then("the URL pathname should be {string}")
    public void urlPathnameShouldBe(String expected) {
        waits().awaitTrue(d -> {
            try {
                return new URI(d.getCurrentUrl()).getPath().equals(expected);
            } catch (Exception e) {
                return false;
            }
        }, waits().defaultWait());
        String actualPath;
        try {
            actualPath = new URI(driver().getCurrentUrl()).getPath();
        } catch (Exception e) {
            actualPath = "(malformed URL)";
        }
        assertEquals(expected, actualPath,
            "URL pathname mismatch — expected '" + expected
            + "' but was '" + actualPath + "'");
    }

    @Then("the SPA should hydrate without console errors")
    public void spaHydratesWithoutConsoleErrors() {
        page().waitForHydration();
        assertNoConsoleSevereErrors();
    }

    // ── Header utility buttons ────────────────────────────────────────────

    @Given("the page theme is {string} or {string}")
    public void captureInitialTheme(String themeA, String themeB) {
        String current = page().currentTheme();
        ScenarioContext.get().setInitialTheme(current);
    }

    @When("I wait {int} milliseconds for the theme transition")
    public void waitForThemeTransition(int ms) {
        try { Thread.sleep(ms); } catch (InterruptedException ie) {
            Thread.currentThread().interrupt();
        }
    }

    @When("I wait {int} milliseconds for the theme to settle")
    public void waitForThemeToSettle(int ms) {
        try { Thread.sleep(ms); } catch (InterruptedException ie) {
            Thread.currentThread().interrupt();
        }
    }

    @Then("the page theme should have changed")
    public void themeShouldHaveChanged() {
        String before = ScenarioContext.get().getInitialTheme();
        String after  = page().currentTheme();
        assertNotEquals(before, after,
            "page theme should have changed but is still '" + after + "'");
    }

    /** Inverse of {@link #themeShouldHaveChanged()} — used by the idempotency
     *  test where a double-toggle should return to the original theme. */
    @Then("the page theme should match the original")
    public void themeShouldMatchOriginal() {
        String before = ScenarioContext.get().getInitialTheme();
        String after  = page().currentTheme();
        assertEquals(before, after,
            "page theme should have returned to the original '" + before
            + "' but is now '" + after + "'");
    }

    @Then("the {word} dropdown should not be visible")
    public void dropdownShouldNotBeVisible(String dropdownName) {
        assertTrue(page().isDropdownClosed(dropdownName),
            "dropdown '" + dropdownName + "' is still visible (expected closed)");
    }

    @Then("a language selection menu should be visible")
    public void languageMenuVisible() {
        assertTrue(page().isLanguageMenuVisible(),
            "language selection menu is not visible after clicking the switcher");
    }

    @Then("the {string} button should be visible and enabled")
    public void buttonVisibleAndEnabled(String label) {
        assertTrue(page().isSignInButtonVisibleAndEnabled(),
            "button '" + label + "' is not visible or not enabled");
    }

    @Then("no console errors should have been logged")
    public void noConsoleErrors() {
        assertNoConsoleSevereErrors();
    }

    // ── Pricing / FAQ assertions ──────────────────────────────────────────

    @Then("I should see the FAQ question {string}")
    public void shouldSeeFaqQuestion(String question) {
        assertTrue(page().isFaqQuestionVisible(question),
            "FAQ question '" + question + "' is not visible");
    }

    // ── Cross-cutting cycle (load runner @health tag) ─────────────────────

    @When("I cycle through every top-level nav item once")
    public void cycleThroughNavItems() {
        String[] navLabels = {"Home", "Pricing", "FAQ", "Contact"};
        RobotActionsHomePage home = page();
        for (String label : navLabels) {
            home.clickNavLink(label);
            // Home nav item resets the URL back to base — a timeout here is
            // tolerated, so use the boolean-returning wait and ignore the result.
            waits().awaitTrue(d -> !RA_HOME.equals(d.getCurrentUrl()), waits().shortWait());
        }
    }

    @Then("the SPA should remain hydrated throughout the cycle")
    public void spaRemainsHydrated() {
        assertTrue(elements().isVisible(By.tagName("main")),
            "<main> is not visible after cycling through all nav items");
    }

    // ── Private helpers ───────────────────────────────────────────────────

    /**
     * Drains the browser log and fails if any SEVERE-level entry is present.
     *
     * Chrome 132+ no longer exposes browser logs via the CDP logging API when
     * the {@code goog:loggingPrefs} capability is absent or when running on
     * certain Grid node configurations.  The call is wrapped in a try/catch so
     * missing log support produces a warning rather than a test failure —
     * infrastructure capability gaps should not mask real assertion failures.
     */
    private void assertNoConsoleSevereErrors() {
        try {
            org.openqa.selenium.logging.LogEntries entries =
                driver().manage().logs().get(
                    org.openqa.selenium.logging.LogType.BROWSER);
            List<org.openqa.selenium.logging.LogEntry> errors =
                entries.getAll().stream()
                    .filter(e -> e.getLevel().equals(java.util.logging.Level.SEVERE))
                    .toList();
            assertTrue(errors.isEmpty(),
                "unexpected console SEVERE errors: " + errors);
        } catch (Exception e) {
            // Browser log retrieval is not supported by this driver/node
            // configuration (common on Chrome 132+).  Log a warning and
            // continue — the absence of log support is not itself an error.
            LOG.warning("browser log retrieval not supported: " + e.getMessage()
                + " — console error check skipped");
        }
    }
}

File: playwright/steps/robotactions-load.steps.ts
/**
 * Step definitions for `features/robotactions-load.feature`.
 *
 * Scope: every nav item on robotactions.com — top-level hash anchors,
 * the Features dropdown (7 items), the Resources dropdown (6 items),
 * header utility buttons, negative cases, deep-link subpages, and the
 * cross-cutting cycle scenario the load runner picks up.
 *
 * Locator strategy follows the playwright-bdd skill:
 *   1. getByRole(role, { name }) wherever possible
 *   2. getByText(text, { exact: true }) for unique visible text
 *   3. Scoped CSS only as fallback (the marketing site has no test-ids)
 */
import { expect } from '@playwright/test';
import { Given, When, Then } from './fixtures';
import type { DataTable } from 'playwright-bdd';

// ── Hero + primary CTAs ────────────────────────────────────────────

// The hero H1 rotates through 3 phrases on ~20s cycle.
// Assert it is non-empty rather than matching a fixed string.
Then('the hero heading should contain one of the known phrases', async ({ page }) => {
  const heading = page.getByRole('heading', { level: 1 }).first();
  await expect(heading).toBeVisible({ timeout: 10_000 });
  const text = (await heading.textContent()) ?? '';
  expect(text.trim().length, 'hero H1 must not be empty').toBeGreaterThan(3);
});

// The stable hero tagline is "No credit card required" — always present
// in the hero regardless of which H1 phrase is currently displayed.
Then('the stable hero tagline should be visible', async ({ page }) => {
  const hero = page.locator('#hero');
  await expect(
    hero.getByText('No credit card required', { exact: false }),
  ).toBeVisible({ timeout: 10_000 });
});

// The hero H1 rotates through 3 phrases on ~20s cycle — assert existence
// and non-empty text rather than a fixed string.
Then('a hero heading should be visible with non-empty text', async ({ page }) => {
  const heading = page.getByRole('heading', { level: 1 }).first();
  await expect(heading).toBeVisible({ timeout: 10_000 });
  const text = (await heading.textContent()) ?? '';
  expect(text.trim().length, 'hero H1 must not be empty').toBeGreaterThan(3);
});

// The hero tagline is stable across all three H1 rotations.
Then('the hero tagline should contain {string}', async ({ page }, fragment: string) => {
  const hero = page.locator('#hero');
  await expect(hero.getByText(fragment, { exact: false })).toBeVisible({ timeout: 10_000 });
});

Then('I should see the {string} button', async ({ page }, label: string) => {
  await expect(page.getByRole('button', { name: label }).first()).toBeVisible();
});

Then('I should see the heading {string}', async ({ page }, label: string) => {
  await expect(
    page.getByRole('heading', { name: label, exact: false }).first(),
  ).toBeVisible({ timeout: 10_000 });
});

// ── FAQ questions ──────────────────────────────────────────────────
Then('I should see the FAQ question {string}', async ({ page }, question: string) => {
  const faq = page.locator('#faq');
  await expect(faq.getByText(question, { exact: false }).first()).toBeVisible({ timeout: 10_000 });
});

// ── Top-level nav anchors ──────────────────────────────────────────
When('I click the {string} nav link', async ({ page }, label: string) => {
  // Nav links may overlap names with footer/CTAs — scope to <nav> first.
  await page.locator('nav').first().getByRole('link', { name: label, exact: true }).click();
});

Then('the URL fragment should be {string}', async ({ page }, fragment: string) => {
  await page.waitForURL(new RegExp(`${fragment.replace('#', '\\#')}$`), { timeout: 5_000 });
});

Then('the section with id {string} should be in viewport', async ({ page }, id: string) => {
  const section = page.locator(`#${id}`).first();
  await expect(section).toBeVisible();
  // Smooth-scroll on the marketing site can take 600-1200ms after the click.
  // Poll the rect until ANY part of the section is inside the viewport,
  // or 5s passes — matches what the user would see after the scroll completes.
  await expect.poll(
    async () =>
      section.evaluate((el) => {
        const r = el.getBoundingClientRect();
        const h = (globalThis as { innerHeight?: number }).innerHeight ?? 0;
        return r.top < h && r.bottom > 0;
      }),
    { timeout: 5_000, message: `section #${id} never entered viewport` },
  ).toBe(true);
});

// ── Dropdowns ───────────────────────────────────────────────────────
When('I click the {string} nav button', async ({ page }, label: string) => {
  await page.locator('nav').first().getByRole('button', { name: label, exact: true }).click();
});

When('I click the {string} button', async ({ page }, label: string) => {
  await page.getByRole('button', { name: label, exact: true }).first().click();
  // Special-case: the "Toggle theme" button opens a menu (Light/Dark/System)
  // rather than flipping the theme directly — verified via DOM snapshot
  // 2026-06-05. Pick the menu item that's the opposite of the current
  // theme so the next "page theme should have changed" assertion has
  // a real state change to observe.
  if (label === 'Toggle theme') {
    const isDark = await page.evaluate(() =>
      document.documentElement.classList.contains('dark'),
    );
    const target = isDark ? 'Light' : 'Dark';
    await page.getByRole('menuitem', { name: target }).first().click();
  }
});

/**
 * Sentinel link that proves a dropdown panel is open.
 *
 * Verified against the live site 2026-08-05: the nav buttons are "Products"
 * and "Resources" — there is no "Features" button. Sentinels must appear ONLY
 * while the panel is open, which rules out "Integrations": that link also sits
 * in the footer, so it reads as visible with the dropdown shut and would let a
 * broken dropdown pass.
 */
const DROPDOWN_SENTINELS: Record<string, string> = {
  Products: 'AI Test Agent',
  Resources: 'Compare',
};

Then('the {word} dropdown should be visible', async ({ page }, dropdownName: string) => {
  // Each dropdown item is the first link to a known route from that menu —
  // assert at least one specific entry is visible to confirm the panel opened.
  const sentinel = DROPDOWN_SENTINELS[dropdownName];
  expect(sentinel, `unknown dropdown ${dropdownName}`).toBeDefined();
  await expect(page.getByRole('link', { name: sentinel, exact: false }).first())
    .toBeVisible({ timeout: 3_000 });
});

Then('the {word} dropdown should not be visible', async ({ page }, dropdownName: string) => {
  const sentinel = DROPDOWN_SENTINELS[dropdownName];
  expect(sentinel, `unknown dropdown ${dropdownName}`).toBeDefined();
  // The dropdown links are still in the DOM when closed (CSS hidden),
  // so we assert they are NOT visible (hidden/display:none).
  await expect(page.getByRole('link', { name: sentinel, exact: false }).first())
    .not.toBeVisible({ timeout: 3_000 });
});

Then(
  'the {word} dropdown should contain the items:',
  async ({ page }, _dropdownName: string, table: DataTable) => {
    // first row of the data table is the header — skip
    const labels = table.rows().map((r) => r[0]);
    for (const label of labels) {
      await expect(
        page.getByRole('link', { name: label, exact: false }).first(),
      ).toBeVisible({ timeout: 3_000 });
    }
  },
);

When('I click the {string} dropdown item', async ({ page }, label: string) => {
  await page.getByRole('link', { name: label, exact: false }).first().click();
});

// Every one of these routes answers 308 -> the same path WITH a trailing
// slash, so the pathname is `/careers` for one moment and `/careers/` the
// next. waitForURL resolves the instant its predicate matches, so a strict
// `=== path` was racing the redirect: it passed when Playwright happened to
// sample before the 308 landed and failed when it sampled after. Three of the
// six Resources rows failed per run, and not the same three.
//
// Normalise both sides so the assertion means "this route", not "this route
// caught at the right millisecond".
Then('the URL pathname should be {string}', async ({ page }, path: string) => {
  const norm = (p: string) => p.replace(/\/+$/, '') || '/';
  await page.waitForURL((url) => norm(new URL(url).pathname) === norm(path), { timeout: 8_000 });
});

Then('the SPA should hydrate without console errors', async ({ page, consoleErrors }) => {
  await page.locator('main').first().waitFor({ state: 'visible', timeout: 10_000 });
  await page.waitForFunction(() => document.readyState === 'complete');
  // Filter third-party tracker noise (same list as the smoke step).
  const ours = consoleErrors.filter((e) => {
    const s = e.toLowerCase();
    if (s.includes('cdn-cgi/rum')) return false;
    if (s.includes('calendly')) return false;
    if (s.includes('cloudflare')) return false;
    if (s.includes('zaraz')) return false;
    if (s.includes('hotjar')) return false;
    return true;
  });
  expect(ours, 'console errors after navigation:\n' + ours.join('\n')).toEqual([]);
});

// ── Header utility buttons ─────────────────────────────────────────
// Module-level variable to track initial theme across a single test's steps.
// playwright-bdd runs each scenario in a fresh worker context so no
// cross-test leakage occurs.
let initialTheme: 'light' | 'dark' | null = null;

Given('the page theme is {string} or {string}', async ({ page }, _a: string, _b: string) => {
  initialTheme = await page.evaluate(() =>
    document.documentElement.classList.contains('dark') ? 'dark' : 'light',
  );
});

When('I wait {int} milliseconds for the theme transition', async ({ page }, ms: number) => {
  await page.waitForTimeout(ms);
});

When('I wait {int} milliseconds for the theme to settle', async ({ page }, ms: number) => {
  await page.waitForTimeout(ms);
});

Then('the page theme should have changed', async ({ page }) => {
  const after = await page.evaluate(() =>
    document.documentElement.classList.contains('dark') ? 'dark' : 'light',
  );
  expect(after, 'theme should flip').not.toBe(initialTheme);
});

Then('the page theme should match the original', async ({ page }) => {
  const after = await page.evaluate(() =>
    document.documentElement.classList.contains('dark') ? 'dark' : 'light',
  );
  expect(after, 'theme should return to original after double toggle').toBe(initialTheme);
});

// Language switcher — the canonical selenium-java feature uses
// "a language selection menu should be visible".
Then('a language selection menu should be visible', async ({ page }) => {
  // The Radix DropdownMenu that opens has role="menu" but no aria-label on the
  // menu container itself. Assert that at least one menuitem (e.g. "English" or
  // "Français") is visible after the button click.
  const menu = page.getByRole('menu').first();
  await expect(menu).toBeVisible({ timeout: 3_000 });
  await expect(menu.getByRole('menuitem').first()).toBeVisible({ timeout: 3_000 });
});

// Kept for backwards compat with the existing playwright-only feature
Then('the language menu should show at least one language option', async ({ page }) => {
  const menu = page.getByRole('menu').first();
  await expect(menu).toBeVisible({ timeout: 3_000 });
  await expect(menu.getByRole('menuitem').first()).toBeVisible({ timeout: 3_000 });
});

Then('the {string} button should be visible and enabled', async ({ page }, label: string) => {
  const btn = page.getByRole('button', { name: label }).first();
  await expect(btn).toBeVisible();
  await expect(btn).toBeEnabled();
});

// ── Navigation to paths (negative + deep-link scenarios) ────────────
When('I navigate to path {string}', async ({ page }, path: string) => {
  await page.goto(path);
});

Then('the page title should contain {string}', async ({ page }, text: string) => {
  await expect(page).toHaveTitle(new RegExp(text, 'i'), { timeout: 10_000 });
});

// ── Cross-cutting cycle (used by the load runner's @health tag) ────
When('I cycle through every top-level nav item once', async ({ page }) => {
  // Anchor links only. "Products" and "Resources" are dropdown BUTTONS, not
  // links — clicking them by role=link never resolves and burns the full
  // action timeout (verified against the live site 2026-08-05).
  const navLabels = ['Home', 'Pricing', 'FAQ', 'Contact'];
  const navScope = page.locator('nav').first();
  for (const label of navLabels) {
    await navScope.getByRole('link', { name: label, exact: true }).click();
    await page.waitForLoadState('networkidle', { timeout: 5_000 }).catch(() => undefined);
  }
});

Then('the SPA should remain hydrated throughout the cycle', async ({ page }) => {
  // Final sanity: <main> still attached and visible after the cycle.
  await expect(page.locator('main').first()).toBeVisible();
});
